tests/autotests/colorcorrection_autotest.cpp

```diff
@@ -0,0 +1,424 @@
+
+    if (connected_displays_.empty())
+      return;
+
+    for (auto &display : connected_displays_)
+      display->SetPowerMode(power_mode);
+  }
+
+  void PopulateConnectedDisplays() {
+    if (connected_displays_.empty()) {
+      connected_displays_ = device_->GetConnectedPhysicalDisplays();
+
+      for (auto &display : connected_displays_) {
+        auto callback = std::make_shared<DisplayVSyncCallback>();
+        display->RegisterVsyncCallback(callback, 0);
+        display->VSyncControl(true);
+      }
+    }
+  }
+
+ private:
+  std::vector<hwcomposer::NativeDisplay *> connected_displays_;
+  hwcomposer::GpuDevice *device_;
+  hwcomposer::SpinLock spin_lock_;
+};
+
+static struct { struct gbm_device *dev; } gbm;
+
+struct drm_fb {
+  struct gbm_bo *bo;
+};
+
+static int init_gbm(int fd) {
+  gbm.dev = gbm_create_device(fd);
+  if (!gbm.dev) {
+    printf("failed to create gbm device\n");
+    return -1;
+  }
+
+  return 0;
+}
+
+static void init_frame(int32_t width, int32_t height) {
+  test_frame.gbm_bo = gbm_bo_create(gbm.dev, width, height, GBM_FORMAT_XRGB8888,
+                                    GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+  if (!test_frame.gbm_bo) {
+    printf("failed to create gbm_bo\n");
+    exit(EXIT_FAILURE);
+  }
+
+  int gbm_bo_fd = gbm_bo_get_fd(test_frame.gbm_bo);
+  if (gbm_bo_fd == -1) {
+    printf("gbm_bo_get_fd() failed\n");
+    exit(EXIT_FAILURE);
+  }
+
+  int planes = gbm_bo_get_num_planes(test_frame.gbm_bo);
+  for (size_t i = 0; i < planes; i++) {
+    test_frame.native_handle.import_data.offsets[i] =
+        gbm_bo_get_plane_offset(test_frame.gbm_bo, i);
+    test_frame.native_handle.import_data.strides[i] =
+        gbm_bo_get_plane_stride(test_frame.gbm_bo, i);
+    test_frame.native_handle.import_data.fds[i] = gbm_bo_fd;
+  }
+
+  test_frame.native_handle.import_data.width = width;
+  test_frame.native_handle.import_data.height = height;
+  test_frame.native_handle.import_data.format =
+      gbm_bo_get_format(test_frame.gbm_bo);
+
+  test_frame.layers_fences.resize(1);
+  hwcomposer::HwcLayer *hwc_layer = NULL;
+  hwc_layer = new hwcomposer::HwcLayer();
+  hwc_layer->SetTransform(0);
+  hwc_layer->SetSourceCrop(hwcomposer::HwcRect<float>(0, 0, width, height));
+  hwc_layer->SetDisplayFrame(hwcomposer::HwcRect<int>(0, 0, width, height));
+  hwc_layer->SetNativeHandle(&test_frame.native_handle);
+  test_frame.layers.push_back(std::unique_ptr<hwcomposer::HwcLayer>(hwc_layer));
+}
+
+void draw_colors(void *ptr, uint32_t height, uint32_t stride, bool gradient) {
+  uint32_t c_offiset = 4;
+  uint32_t c_value = 255;
+
+  memset(ptr, 0, height * stride);
+  for (int i = 0; i < height; i++) {
+    if (i < height / 3) {
+      c_offiset = 0;  // blue
+    } else if (i >= height / 3 && i < (height * 2) / 3) {
+      c_offiset = 1;  // green
+    } else {
+      c_offiset = 2;  // red
+    }
+    if (gradient) {
+      c_value = 255 * ((float)(i % (height / 3)) / (height / 3));
```

Comment:
we might want to have a check here that height/3 > 0 or modulo result will be undefined. i.e store height / 3 in a temp variable, ensure its > 0 and than do this calculation